### PR TITLE
Should SA1307 check verbatim identifiers?

### PR DESCRIPTION
**Pull request by NextTurn**

### CODE REVIEW

**Comment by NextTurn** [comment]
Yes, it did fail.

https://ci.appveyor.com/project/sharwell/stylecopanalyzers/builds/28919339

File: StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/RenameToUpperCaseCodeFixProvider.cs [hunk 1]
```diff
@@ -55,7 +55,7 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
             foreach (var diagnostic in context.Diagnostics)
             {
                 var token = root.FindToken(diagnostic.Location.SourceSpan.Start);
-                var tokenText = token.ValueText.TrimStart('_');
+                var tokenText = token.ValueText.TrimStart('_').TrimStart('@');
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers.CodeFixes/NamingRules/RenameToUpperCaseCodeFixProvider.cs]: :memo: ValueText doesn't have the `@` prefix, right?

File: StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1307AccessibleFieldsMustBeginWithUpperCaseLetter.cs [hunk 2]
```diff
@@ -76,7 +76,7 @@ private static void HandleFieldDeclaration(SyntaxNodeAnalysisContext context)
                         string name = declarator.Identifier.ToString();
```
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1307AccessibleFieldsMustBeginWithUpperCaseLetter.cs]: :thought_balloon: maybe this should just be using ValueText?

**Comment by NextTurn** [comment]
I got some errors when building locally:

```
error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly. [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018: LibGit2Sharp.NotFoundException: object not found - no match for id (68156634b00fe68a631c09add8bab17faf2e7db8) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.Core.Proxy.git_revwalk_next(RevWalkerHandle walker) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.CommitLog.CommitEnumerator.MoveNext() [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at Nerdbank.GitVersioning.VersionOracle..ctor(String projectDirectory, Repository repo, Commit head, ICloudBuild cloudBuild, Nullable`1 overrideBuildNumberOffset, String projectPathRelativeToGitRepoRoot) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at Nerdbank.GitVersioning.VersionOracle.Create(String projectDirectory, String gitRepoDirectory, ICloudBuild cloudBuild, Nullable`1 overrideBuildNumberOffset, String projectPathRelativeToGitRepoRoot) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner() [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018: The "Nerdbank.GitVersioning.Tasks.GetBuildVersion" task failed unexpectedly. [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018: LibGit2Sharp.NotFoundException: object not found - no match for id (68156634b00fe68a631c09add8bab17faf2e7db8) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.Core.Proxy.git_revwalk_next(RevWalkerHandle walker) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at LibGit2Sharp.CommitLog.CommitEnumerator.MoveNext() [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at Nerdbank.GitVersioning.VersionOracle..ctor(String projectDirectory, Repository repo, Commit head, ICloudBuild cloudBuild, Nullable`1 overrideBuildNumberOffset, String projectPathRelativeToGitRepoRoot) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at Nerdbank.GitVersioning.VersionOracle.Create(String projectDirectory, String gitRepoDirectory, ICloudBuild cloudBuild, Nullable`1 overrideBuildNumberOffset, String projectPathRelativeToGitRepoRoot) [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
error MSB4018:    at Nerdbank.GitVersioning.Tasks.GetBuildVersion.ExecuteInner() [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers\StyleCop.Analyzers.csproj]
```

When building *StyleCop.Analyzers.CodeFixes* for .NET Standard 1.1:

```
C:\Program Files\dotnet\sdk\3.1.100-preview3-014645\Sdks\NuGet.Build.Tasks.Pack\buildCrossTargeting\NuGet.Build.Tasks.Pack.targets(198,5): error : Id is required. [StyleCopAnalyzers\StyleCop.Analyzers\StyleCop.Analyzers.CodeFixes\StyleCop.Analyzers.CodeFixes.csproj]
```

Why do we need that *global.json*? I only have .NET Core 3.1 SDK and it can build well.

**Comment by sharwell** [comment]
> LibGit2Sharp.NotFoundException: object not found - no match for id

Did you use a shallow clone? The build relies on a full clone.

> Why do we need that global.json? I only have .NET Core 3.1 SDK and it can build well.

Warnings as errors are turned on in the build. Different versions of the SDK produced different warnings, so the global.json was required for stabilizing the build.

**Comment by NextTurn** [comment]
> > LibGit2Sharp.NotFoundException: object not found - no match for id
> 
> Did you use a shallow clone? The build relies on a full clone.

Yes, I always use shallow clones to save time and space.

**Comment by sharwell** [comment]
> Yes, I always use shallow clones to save time and space.

See AArnott/Nerdbank.GitVersioning#67 and AArnott/Nerdbank.GitVersioning#92 for background here. This project will only be able to support builds from shallow clones if the dependency does as well.

**Comment by sharwell** [comment]
Thanks @NextTurn !

**Pull request merged**